### ISSUE TITLE
`textDocument/definition`

### ISSUE THREAD

**Issue by dpetka2001** [abeldekat/nvim_pde#1]
Hi Chris, I hope you're doing well.

https://github.com/abeldekat/nvim_pde/blob/3192b5a87e191186bb5bb22d787a7934b342024e/lua/ak/config/lang/lspconfig.lua#L90-L96

This seems to not be working any more (not sure when it stopped working tbh). According to https://github.com/neovim/neovim/blob/8260e4860b27a54a061bd8e2a9da23069993953a/runtime/doc/news.txt#L82-L85 and also https://github.com/neovim/neovim/pull/30982 the `textDocument/definition` doesn't trigger the `vim.lsp.handlers` that we configure in the `nvim-lspconfig` setup of a server. 

A dirty workaround (I believe there has to be a better way) is this autocmd which modifies directly the request from the client
```lua
vim.api.nvim_create_autocmd("LspAttach", {
  callback = function(args)
    local client = vim.lsp.get_client_by_id(args.data.client_id)
    if client and client.name == "lua_ls" then
      -- Store the original request method
      local orig_request = client.request
      -- Override the request method
      client.request = function(method, params, handler, bufnr)
        if method == "textDocument/definition" then
          -- Override the handler of the request with our custom one
          -- to modify `result`
          return orig_request(method, params, function(err, result, ctx, config)
            if type(result) == "table" and result and #result > 0 then
              -- Force single result before passing to handler
              result = { result[1] }
            end
            return handler(err, result, ctx, config)
          end, bufnr)
        end
        -- For the rest of methods return original request with no custom handler
        return orig_request(method, params, handler, bufnr)
      end
    end
  end,
})
```
Maybe there is some other way to do this more cleanly in the `nvim-lspconfig` setup but I'm not aware of it.

**Comment by dpetka2001** [comment]
A more concise solution would be to create a keymap like the following
```lua
vim.keymap.set("n", "<C-q>", function()
  local params = vim.lsp.util.make_position_params(0, "utf-16")
  vim.lsp.buf_request(0, "textDocument/definition", params, function(_, result, _, _)
    if type(result) == "table" and result and #result > 0 then
      -- Force single result before passing to handler
      result = { result[1] }
      vim.cmd("FzfLua lsp_definitions jump_to_single_result=false")
    end
  end)
end, { desc = "Lsp Definition" })
```
Use whatever fuzzy finder picker you would like. Looks to me much better than the autocmd. Inspiration taken from `noice.nvim` which overrides the `hover`/`signature` handlers :smile: 

Edit: Nope, not working. I still had the autocmd active and didn't realize it. Sorry for noise.

**Comment by abeldekat** [comment]
Thanks!

At the moment I am not using nightly.
I added a note to handle this when version 0.11 is released.

Echasnovski already refactored his [personal config](https://github.com/echasnovski/nvim/commit/f30b003cb2f30b928a03a095af849fd49c2c95e9)

**Comment by dpetka2001** [comment]
It seems that the autocmd is not needed in my case. fzf-lua provides an option `unique_line_items`, which does the same thing (it has its own custom location handler for the LSP providers). So, I just enabled that option in my keymaps :smile: 

The autocmd I mentioned is a bit buggy, because there are instances where there are multiple results but on different files and line numbers (in the LazyVim codebase that I tested) and it only returns the first one which is not what you want. 

So, the way Echasnovski did it, by tracking the filename and line number is more correct (and FzfLua does the same thing internally with that option).
